### PR TITLE
Ship the reporter as a package any web app can install

### PR DESCRIPTION
**Pull request by hotsunchip**
## Summary

Third of four, and the point where the prototype is complete: the reporter sees
the agent's issue text and its proposed PR grouping, edits both, and confirms.
**Nothing is filed.** Base: `feat/debug-report-capture`.

Drafting happens at **preview** time rather than after handover, which is what
makes the confirmation meaningful — approving a batch should mean seeing what
will be filed, not authorising a translation you have not read. The panel is
also the privacy gate, and a stronger one than any server-side rule: a human
looks at each image before it leaves.

**The drafting call is tool-free by construction.** There is no `tools`
parameter on the request, so the session cannot read a file, reach the
repository or open a pull request; the worst case is a wasted token budget and a
draft the reporter rejects. That absence is what makes it acceptable for the
drafting credential to sit with the app while the pipeline credential — the one
that can create commits — stays with the repository. The key is read in the
dev-server process and never reaches the browser.

Grouping is proposed by **kind, not count**: two spacing fixes pass or fail
review together, a spacing fix and a behaviour fix do not. `layout` and `logic`
never group. The reporter can detach an item, split a PR or merge two; merging
across kinds warns rather than blocks, because the person is the author of
record.

With no credential the path **degrades** instead of failing — the reporter's own
sentences travel, one PR per item, and the panel says so.

### Deviation from the plan

It was going to route through `scripts/agent/ask.mjs`, which requires at least
one read tool ("an agent that cannot read cannot cite evidence"). Correct for a
verifier; wrong for a call whose whole argument is having no tools. The Messages
API is called directly rather than widening a shared security module.

## Test plan

- `pnpm verify:fast` green
- Unit: the draft schema and its fail-closed parser, grouping-rule
  normalisation, detach/split/merge invariants, the session PR cap, the
  no-credential branch, and both dev endpoints' guards
- Both endpoints require same-origin **and** `application/json`, so a
  cross-origin page cannot drive a port that writes into the repository and
  spends a model credential. `.` and `..` are rejected as path segments —
  `path.join(root, ".wb-reports", "..")` is the repository root
- Manual smoke in `pnpm dev`, and a real batch taken through the panel

## Stack

1. `feat/debug-report`
2. `feat/debug-report-capture`
3. **this PR** — preview, drafting, grouping
4. `feat/debug-report-intake`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added a development debug-report workflow for reviewing, editing, grouping, and submitting captured issues.
  * Added AI-assisted draft summaries and suggested homogeneous groups, with fallback handling when drafting is unavailable.
  * Added capture previews, notes, issue dispositions, candidate flags, and controls to detach, split, or merge groups.
  * Added a review panel accessible from the badge or with `v`; Escape closes the panel before changing debug mode.
  * Added report metadata, image handling, session limits, queued-item preservation, and clear handoff status messages.

* **Documentation**
  * Documented rollout stages, authorization boundaries, and deferred deployed-environment support.

### CODE REVIEW

File: packages/frontend/src/debug/host.ts [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+      error: `the dev server did not answer (${err instanceof Error ? err.message : String(err)})`,
+    };
+  } finally {
+    if (timer !== undefined) clearTimeout(timer);
+  }
+  let data: unknown;
+  try {
+    data = await response.json();
+  } catch {
+    data = undefined;
```
[the remark this answers is not in the document]
Comment (hotsunchip) [reply packages/frontend/src/debug/host.ts]: Fixed in 0d9beb88e. `fetch` resolves once the headers arrive, so `clearTimeout` in that `finally` left `response.json()` with no deadline at all — and an unsendable batch is precisely what `DRAFT_TIMEOUT_MS` exists to prevent, so the timeout was absent from the half where it mattered.

The timer now covers the body, and `abort()` rejects the body read too, which is why one timer is enough for both halves. A body that never arrived and a body that is not JSON now report differently — the first is the timeout doing its job and has to say so, rather than degrading into "the response was not JSON".

File: packages/debug-report/src/ui/handover.ts [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+}
+
+/** One line the panel shows above the button. */
+export function handoverSummary(
+  items: readonly DebugItem[],
+  groups: readonly ProposedGroup[],
+): string {
+  const summary = summariseBundle(items, groups);
+  const { queued } = withinSessionCap(groups);
+  const parts = [
```
[the remark this answers is not in the document]
Comment (hotsunchip) [reply packages/debug-report/src/ui/handover.ts:227]: Fixed in 0d9beb88e. `handoverSummary` capped the raw proposal while `handOver` capped the live groups, so the panel could report a PR as waiting for the next batch while the handover sent it.

Rather than repeat the filter, both now call one `liveGroups(items, groups)` — two copies of "which groups are still alive" is exactly how they came to disagree, and a third copy in a test would have hidden it again. The boundary case you asked for is covered: five live groups plus one naming only discarded items, asserting nothing is reported as waiting.

File: packages/debug-report/src/ui/overlay.tsx [hunk 3]
```diff
@@ -201,7 +185,9 @@ export function DebugOverlay({ route }: { route: string }) {
         isTypingTarget(e.target);
       if (inField) return;
 
-      const action = actionFor(e, live);
+      // While the panel is open only leaving it is recognised, so its buttons
+      // keep their own keyboard activation and nothing underneath is aimed at.
+      const action = reviewing ? actionWhileReviewing(e) : actionFor(e, live);
```
[the remark this answers is not in the document]
Comment (hotsunchip) [reply packages/debug-report/src/ui/overlay.tsx:205]: Fixed in 0d9beb88e. `reviewing` is view-local, so it outlived the mode: turning debug mode off with the panel open left it set, and the next toggle reopened the panel without the reporter asking. The mode is the outer state and the panel cannot outlive it.

Test walks the whole path — open the panel, toggle off, toggle on — and asserts the badge is back while the panel is not.

File: packages/frontend/src/debug/session-id.ts [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+ * ONE derivation, imported by both users. It was computed independently in
+ * `mount.tsx` (the `.wb-reports/` directory name) and in `use-debug-session.ts`
+ * (the `localStorage` key), so the report directory and the persisted session
+ * agreed only because both modules happened to evaluate in the same millisecond
+ * — and neither file mentioned the other.
+ *
+ * Derived from the clock rather than randomly so a directory listing sorts the
+ * way the sessions happened.
+ */
+export const DEBUG_SESSION_ID = `wb-${Date.now().toString(36)}`;
```
[the remark this answers is not in the document]
Comment (hotsunchip) [reply packages/frontend/src/debug/session-id.ts]: Fixed in 0d9beb88e. Two tabs opened in the same millisecond took the same id, and with it the same `.wb-reports/<sessionId>` directory and the same persistence key — so one tab's bundle could be written over the other's, and the store's adoption check read both as one session.

Built through an injectable factory as you suggested, so the collision is a test rather than an argument: pin the clock, vary only the randomness, assert the ids differ.

Writing that test corrected the docblock too. It claimed the clock prefix keeps a directory listing sorted, and a lexicographic sort only agrees with a numeric one at a FIXED WIDTH — my first test used 1_000 and 2_000, which are two and three base-36 characters and sort the other way. Measured: real epoch milliseconds are eight base-36 characters from 1975 to 2059, so the property is real but not eternal, and the comment now says so.

File: packages/frontend/src/debug/appearance.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+/** One line naming what a report is about. */
+export function describeTarget(target: Target): string {
+  if (target.kind === "dom") {
+    return `${target.tag}${target.testId ? ` · ${target.testId}` : ""} · ${target.selector}`;
+  }
+  if (target.kind === "canvas") {
+    return `${target.surface}${target.address ? ` · ${target.address}` : " · no address"}`;
+  }
+  return `region${target.elements ? ` · ${target.elements.length} element(s)` : ""}`;
```
[the remark this answers is not in the document]
Comment (hotsunchip) [reply packages/frontend/src/debug/appearance.ts]: Partly applied in 0d9beb88e, and I want to be explicit about which half.

Agreed on the structure: the `viewport` branch was the FALLBACK, so a kind added later would have quietly described itself as a region. It is now explicit, and an unrecognised kind returns `unknown target (<kind>)` instead of borrowing wording that does not belong to it.

Not changing the word: `region` is deliberate. `viewport` is the internal kind name for a dragged rectangle, and "region" is what the reporter is shown everywhere else — the badge, the note form, the panel. Showing `viewport` there would make the panel and the overlay describe one thing two ways, which `appearance.ts` exists to prevent.

Tests cover all four: dom, canvas with and without an address, viewport with and without elements, and an unknown kind.

**Comment by hotsunchip** [comment]
Heads-up before the next review pass: this branch was force-pushed with a **structural change**, so the diff has moved under the five threads above. Their fixes are all still in — the paths changed.

The overlay, the panel, the capture and the transport were in `packages/frontend/src/debug/`, which made them part of this application. They are now `@wafflebase/debug-report/react`, with `/plugin` for the two dev-server endpoints and `/testing` for host-wiring helpers. React is a peer dependency of `/react` alone, so the core entry stays free of it.

The one line that mattered: `locate.ts` imported the sheet and doc locators directly, and that is what tied the whole overlay to this repository. The canvas branch is injected now (`locateOnCanvas`), and a host that omits it gets a region for every canvas point — the honest answer for a surface nothing outside it can interrogate.

`packages/frontend/src/debug/` is down to what is genuinely wafflebase: the two engine locators, the surface registry they answer through, the route patterns, and the mount.

Doing this before merge rather than after was deliberate — the next two PRs in the stack build on these paths, so moving later would mean a second pass over the same files.

Verified: package typecheck clean, 283 package tests, 21 frontend debug tests, lint, both frontend tsconfigs, `verify:entropy`.

**Comment by hotsunchip** [comment]
@coderabbitai review

Requesting a fresh pass: this branch was force-pushed with a structural change and the incremental diff does not represent it. About 3,000 lines moved from `packages/frontend/src/debug/` into `@wafflebase/debug-report` as three new entry points (`/react`, `/plugin`, `/testing`), so the files under the five resolved threads above are at new paths.

Worth reviewing with fresh eyes rather than as a delta:

- the new package boundary — whether anything left in `packages/frontend/src/debug/` is still generic, or anything moved is still wafflebase-specific
- `locateOnCanvas`, the seam that replaced a direct import of the engine locators. A host that omits it gets a region for every canvas point
- `peerDependencies` on the package: `react`/`react-dom`/`vite`, all optional, since the core entry needs none of them

The title and commit subject changed to match what the PR now is.

**Comment by hotsunchip** [comment]
@claude review

Requesting a fresh pass: this branch was force-pushed with a structural change and the incremental diff does not represent it. About 3,000 lines moved from packages/frontend/src/debug/ into @wafflebase/debug-report as three new entry points (/react, /plugin, /testing), so the files under the five resolved threads above are at new paths.

Worth reviewing with fresh eyes rather than as a delta:

the new package boundary — whether anything left in packages/frontend/src/debug/ is still generic, or anything moved is still wafflebase-specific
locateOnCanvas, the seam that replaced a direct import of the engine locators. A host that omits it gets a region for every canvas point
peerDependencies on the package: react/react-dom/vite, all optional, since the core entry needs none of them
The title and commit subject changed to match what the PR now is.

**Review by hackerwins — APPROVED** [review]
Thanks for your contribution.
